src/Common/src/System/Globalization/IdnMapping.Unix.cs

```diff
@@ -36,25 +54,43 @@ private unsafe string GetAsciiCore(char* unicode, int count)
         private unsafe string GetUnicodeCore(char* ascii, int count)
         {
             uint flags = Flags;
-            CheckInvalidIdnCharacters(ascii, count, flags, "ascii");
+            CheckInvalidIdnCharacters(ascii, count, flags, nameof(ascii));
 
-            char[] buf = new char[count];
-
-            for (int attempts = 2; attempts > 0; attempts--)
+            const int StackAllocThreshold = 512;
+            if (count < StackAllocThreshold)
             {
-                int realLen = Interop.GlobalizationNative.ToUnicode(flags, ascii, count, buf, buf.Length);
-
-                if (realLen == 0)
+                char* output = stackalloc char[count];
+                return GetUnicodeCore(ascii, count, flags, output, count, reattempt: true);
+            }
+            else
+            {
+                char[] output = new char[count];
+                fixed (char* pOutput = output)
                 {
-                    break;
+                    return GetUnicodeCore(ascii, count, flags, pOutput, count, reattempt: true);
                 }
+            }
+        }
+
+        private unsafe string GetUnicodeCore(char* ascii, int count, uint flags, char* output, int outputLength, bool reattempt)
+        {
+            int realLen = Interop.GlobalizationNative.ToUnicode(flags, ascii, count, output, outputLength);
 
-                if (realLen <= buf.Length)
+            if (realLen == 0)
+            {
+                throw new ArgumentException(SR.Argument_IdnIllegalName, nameof(ascii));
```

Comment:
actually I asked for the opposite in the first place. I believe @hughbe  had it as you suggested. my only point is I don't like to have a block of code inside parenthesizes while we can just return or throw and keep the code flat fo rthe sake of readability. I can see the duplication of throw but I don't think this is big deal.